[dynamo] allow tensor subclass fakification in dynamo

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #103146
* __->__ #105308

This PR adds necessary plumbing through torchdynamo to allow tensor
subclasses with certain contract (i.e. with `__tensor_flatten__` and
`__tensor_unflatten__`) to goes through the dynamo fakification pass by
fakifying the tensor subclass internal components.

Some of the tensor subclass contract logic mostly borrowed from
https://github.com/pytorch/pytorch/pull/97540

Added some tests to verify simply passing through a tensor subclass
(i.e. DTensor) through dynamo eager works as expected.

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

wanchaol committed:
[dynamo] allow tensor subclass fakification in dynamo

This PR adds necessary plumbing through torchdynamo to allow tensor
subclasses with certain contract (i.e. with `__tensor_flatten__` and
`__tensor_unflatten__`) to goes through the dynamo fakification pass by
fakifying the tensor subclass internal components.

Some of the tensor subclass contract logic mostly borrowed from
https://github.com/pytorch/pytorch/pull/97540

Added some tests to verify simply passing through a tensor subclass
(i.e. DTensor) through dynamo eager works as expected.

[ghstack-poisoned]

awgu commented on torch/distributed/_tensor/api.py:
```diff
@@ -213,6 +213,26 @@ def __repr__(self):
+
+    @staticmethod
+    def __tensor_unflatten__(local_tensor, spec):
+        assert spec is not None
```
nit: If this assertion is raised, should we point the user to `self._spec` being `None` in the `__tensor_flatten__()` return value?

wanchaol committed:
Update on "[dynamo] allow tensor subclass fakification in dynamo"


This PR adds necessary plumbing through torchdynamo to allow tensor
subclasses with certain contract (i.e. with `__tensor_flatten__` and
`__tensor_unflatten__`) to goes through the dynamo fakification pass by
fakifying the tensor subclass internal components.

Some of the tensor subclass contract logic mostly borrowed from
https://github.com/pytorch/pytorch/pull/97540

Added some tests to verify simply passing through a tensor subclass
(i.e. DTensor) through dynamo eager works as expected.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

voznesenskym commented on torch/distributed/_tensor/api.py:
```diff
@@ -213,6 +213,28 @@ def __repr__(self):
         # TODO: consider all_gather the local tensors for better debugging
         return f"DTensor(local_tensor={self._local_tensor}, device_mesh={self._spec.mesh}, placements={self._spec.placements})"
 
+    def __tensor_flatten__(self):
```
I am not 100% sold on this being a protocol/API we want to exposed versus just doing this behavior for DTensor only. Thoughts?

wanchaol replied:
I think we can definitely do this behavior only for DTensor by special casing DTensor fakification in dynamo, but I think for other tensor subclasses this might be useful as long as they defined the similar interface, i.e. @bdhirsh is also trying to enable `parseSemiStructuredTensor` which also needs to define `__tensor_flatten__` and `__tensor_unflatten__` https://fburl.com/zz802okh, so I feel exposing this protocol might makes more uniform?

ezyang replied:
I'm definitely pro having a generalized APIs; there are more subclasses than just DTensor. If we need to bikeshed the protocol, we can, but I think we should have *something*.

wanchaol committed:
Update on "[dynamo] allow tensor subclass fakification in dynamo"


This PR adds necessary plumbing through torchdynamo to allow tensor
subclasses with certain contract (i.e. with `__tensor_flatten__` and
`__tensor_unflatten__`) to goes through the dynamo fakification pass by
fakifying the tensor subclass internal components.

Some of the tensor subclass contract logic mostly borrowed from
https://github.com/pytorch/pytorch/pull/97540

Added some tests to verify simply passing through a tensor subclass
(i.e. DTensor) through dynamo eager works as expected.

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

ezyang commented:
@pytorchbot merge
